### PR TITLE
Starting work on negation, wip

### PR DESCRIPTION
**Pull request by ceymard**
Here is a naive interpretation on how the negation could function.

If you want to do it entirely yourself, go ahead, but I'm willing to see this through as I really would like to see it come to fruition and this helps me hone my go funk.

I'm looking for some feedback on that preliminary implementation.

Am I missing something glaring ? Is the clone stuff potentially a performance killer ? Should Parse() be a little more involved ?

### CODE REVIEW

**Comment by ceymard** [comment]
I could change all that to have ` !( ';' | ',' | SomeToken)` by parsing it by hand (not going for parseDisjunction) and having a Parse() that actually peeks and verifies the list passes the check. That would ensure that we forbid anything like `!( @'what' )` which would make no sense since for binding we'd rather have `@!'lit'` or `@!(';', ',')`

**Comment by alecthomas** [comment]
Yes, as you suspect, clone is slow.

> I could change all that to have !( ';' | ',' | SomeToken) by parsing it by hand (not going for parseDisjunction) and having a Parse() that actually peeks and verifies the list passes the check.

It would be simpler to just call `parseDisjunction()` then walk it to type check it. But I think just supporting `![@]<literal>` or `![@]<reference>` to start with is a good start, then we could expand it to

File: grammar.go [hunk 1]
```diff
@@ -282,6 +284,27 @@ func (g *generatorContext) parseGroup(slexer *structLexer) (node, error) {
...
+	// Now, verify that next is either a literal or a token
+	// literal / reference
+	switch next.(type) {
+	case *literal:
+		return &negation{next}, nil
+	case *reference:
+		return &negation{next}, nil
+	default:
+	}
+	return nil, fmt.Errorf("Negations can only handle literals or single tokens")
```
Comment (alecthomas) [inline grammar.go:305]: No capitalisation in errors please, and this should have some context like `fmt.Errorf("only literals or references are negatable, not %s", next)

File: nodes.go [hunk 2]
```diff
@@ -412,6 +412,31 @@ func (l *literal) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.
...
 
+type negation struct {
+	node node
+}
+
+func (n *negation) String() string { return "!" + stringer(n.node) }
+
+func (n *negation) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.Value, err error) {
+	// We create a clone to go back to the previous lexer
+	clone := newParseContext(ctx.Clone(), ctx.lookahead, ctx.caseInsensitive)
```
Comment (alecthomas) [inline nodes.go:423]: Use `ctx.Branch()` and `ctx.Accept()`/`ctx.Stop()` for this conditional branching. See [optional](https://github.com/alecthomas/participle/blob/master/nodes.go#L330-L345) for an example.

**Review by alecthomas — CHANGES_REQUESTED** [review]
Good start!

You'll also need to update the visitors.

**Comment by alecthomas** [comment]
BTW, if you want to join my [channel](https://gophers.slack.com/archives/CN9DS8YF3) in the Gophers Slack, there's a little group of us working on Participle :)

**Comment by alecthomas** [comment]
Thanks for this by the way, looking good!

**Comment by ceymard** [comment]
Alright.

So I added some tests, which show so far that the negation has the potential to be pretty flexible.

It seems just parsing a term (with groups, sequences and whatnot) without restrictions has some satisfying results, even when I want to negate sequences (which is probably useful). Are there potential restrictions to address ?

**However**, I'm ignoring lookahead by not calling `Stop()` in `negation.Parse`. I did that because of course, when negating a sequence, I need the lookahead to be at least as big as the potential sequence, but it works if we just ignore that (so that Stop won't Accept my arm if partially matched, thus failing to parse afterwards). Am I missing something ? Is the lookahead size important to prevent the parser from going haywire or something ?

Also, I'm a bit puzzled by the depth in the stringer(). What is its objective ?

**Comment by ceymard** [comment]
I reverted my tentative changes in stringer(), but the thing is if I define `!('tok' 'tok2')`, then stringer just won't display 'tok2' since the sequence stops at depth 0 which is attained immediately (so right after the first item).

A '...' would be nice, or even better have everything on display. I'm just not sure how depth is supposed to work.

**Comment by ceymard** [comment]
> BTW, if you want to join my [channel](https://gophers.slack.com/archives/CN9DS8YF3) in the Gophers Slack, there's a little group of us working on Participle :)

How do I join ? No matter how I try to login, I just can't enter. :(

**Comment by alecthomas** [comment]
> > BTW, if you want to join my [channel](https://gophers.slack.com/archives/CN9DS8YF3) in the Gophers Slack, there's a little group of us working on Participle :)
> 
> How do I join ? No matter how I try to login, I just can't enter. :(

Try this link I think? https://invite.slack.golangbridge.org/

I'm in Australia, so I'll be back online in about 9-10 hours.

**Comment by alecthomas** [comment]
> It seems just parsing a term (with groups, sequences and whatnot) without restrictions has some satisfying results, even when I want to negate sequences (which is probably useful). Are there potential restrictions to address ?

No I think the approach is sound.

> **However**, I'm ignoring lookahead by not calling `Stop()` in `negation.Parse`. I did that because of course, when negating a sequence, I need the lookahead to be at least as big as the potential sequence, but it works if we just ignore that (so that Stop won't Accept my arm if partially matched, thus failing to parse afterwards).

Sounds reasonable.

> Also, I'm a bit puzzled by the depth in the stringer(). What is its objective ?

It's not the recursion depth. The idea is that it displays to the user the next set of options available to them, which is why in the case of `"foo" "bar"` only `"foo"` would be displayed.

File: grammar.go [hunk 3]
```diff
@@ -282,6 +284,20 @@ func (g *generatorContext) parseGroup(slexer *structLexer) (node, error) {
 	return &group{expr: disj}, nil
 }
 
+// A token negation
+//
+// Accepts both the form !"some-literal" and !SomeNamedToken
+func (g *generatorContext) parseNegation(slexer *structLexer) (node, error) {
+	_, _ = slexer.Next() // advance the parser since we have '!' right now.
+	next, err := g.parseTermNoModifiers(slexer)
+	// TODO: maybe have a warning of some kind when somewhere we have a node
```
Comment (alecthomas) [inline grammar.go:293]: You can remove this TODO - that's exactly what #107 is for.

File: nodes.go [hunk 4]
```diff
@@ -412,6 +412,44 @@ func (l *literal) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.
...
+
+func (n *negation) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.Value, err error) {
+	// Create a branch to avoid advancing the parser, but call neither Stop nor Accept on it
+	// since we will discard a match.
+	branch := ctx.Branch()
+	notEOF, err := ctx.Peek(0)
+	if err != nil {
+		return nil, err
+	}
+	if notEOF.Type == lexer.EOF {
```
Comment (alecthomas) [inline nodes.go:429]: Use `token.EOF()` instead.

File: nodes.go [hunk 5]
```diff
@@ -412,6 +412,44 @@ func (l *literal) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.
...
+		// EOF cannot match a negation, which expects something
+		return nil, nil
+	}
+
+	out, err = n.node.Parse(branch, parent)
+
+	if out != nil && err == nil {
+		// DOUBT : I'm not using stop because it has a possibility of accepting the branch
```
Comment (alecthomas) [inline nodes.go:437]: I think your assessment is correct, just nuke these comments.

File: parser_test.go [hunk 6]
```diff
@@ -1212,3 +1212,59 @@ func TestPointerToList(t *testing.T) {
...
+	require.Error(t, err)
+}
+
+func TestNegationWithPattern(t *testing.T) {
+	type grammar struct {
+		EverythingMoreComplex *[]string `@!(';' String)* @';' @String`
+	}
+
+	// Note: we need more lookahead since (';' String) needs some before failing to match
+	//         this is only if we use Stop() which may Accept if we're past the lookahead count...
```
Comment (alecthomas) [inline parser_test.go:1236]: `UseLookahead(2)`?

File: parser_test.go [hunk 7]
```diff
@@ -1212,3 +1212,59 @@ func TestPointerToList(t *testing.T) {
...
+	err = p.ParseString(`hello ; world ;`, ast)
+	require.Error(t, err)
+}
+
+func TestNegationWithDisjunction(t *testing.T) {
+	type grammar struct {
+		EverythingMoreComplex *[]string `@!(';' | ',')* @(';' | ',')`
+	}
+
+	// Note: we need more lookahead since (';' String) needs some before failing to match
```
Comment (alecthomas) [inline parser_test.go:1259]: Comment was copy and pasted?

File: stringer.go [hunk 8]
```diff
@@ -90,6 +97,13 @@ func (s *stringerVisitor) visit(n node, depth int) { // nolint: gocognit
 		}
 		fmt.Fprint(s, "*")
 
+	case *negation:
+		fmt.Fprintf(s, "!")
+		fmt.Fprint(s, "(")
```
Comment (alecthomas) [inline stringer.go:102]: Parentheses here and below should be guarded, as the other cases are, with:

```go
composite := compositeNode(map[node]bool{}, n)
if composite {
  fmt.Fprint(s, "(")
}
```

File: stringer.go [hunk 9]
```diff
@@ -169,6 +183,7 @@ func compositeNode(seen map[node]bool, n node) bool {
 		return compositeNode(seen, n.expr)
 
 	default:
+		log.Printf("%T", n)
```
Comment (alecthomas) [inline stringer.go:186]: Remove this.

File: stringer.go [hunk 10]
```diff
@@ -150,7 +164,7 @@ func compositeNode(seen map[node]bool, n node) bool {
 		}
 		return false
 
-	case *reference, *literal, *parseable:
+	case *reference, *literal, *parseable, *negation:
```
Comment (alecthomas) [inline stringer.go:167]: Negation should be its own case and recurse into its sub-expression. See eg. `*strct`.

**Review by alecthomas — CHANGES_REQUESTED** [review]
Seems pretty close. I'm happy to merge once my comments are addressed. Thanks!

File: parser_test.go (hunk 7 above)
Comment (ceymard) [reply parser_test.go:1259]: Yes...

File: parser_test.go (hunk 6 above)
Comment (ceymard) [reply parser_test.go:1236]: No need for that anymore actually since I ignore Stop !

File: grammar.go [hunk 11]
```diff
@@ -282,6 +284,18 @@ func (g *generatorContext) parseGroup(slexer *structLexer) (node, error) {
 	return &group{expr: disj}, nil
 }
 
+// A token negation
+//
+// Accepts both the form !"some-literal" and !SomeNamedToken
```
Comment (alecthomas) [inline grammar.go:289]: Is this comment accurate now?

**Comment by alecthomas** [comment]
Awesome, thanks!

If you want to `git rebase -i master && git push -f`, we can use this PR. Up to you.

**Comment by ceymard** [comment]
Done !

**Comment by alecthomas** [comment]
Thanks again :)

**Comment by ceymard** [comment]
fixes #104 for reference.

**Pull request merged**